File: lib/prompts/templates/stepUp.ts
import type { Blueprint, GenerateRequest, SectionOutput, SectionSpec } from '../../types'

export function buildPrompt(
  spec: SectionSpec,
  blueprint: Blueprint,
  req: GenerateRequest,
  priorOutputs?: SectionOutput[]
): string {
  const { lesson_metadata } = req

  const assessmentPassageOutput = priorOutputs?.find(
    (o) => o.section_type === 'assessment_passage'
  )
  const assessmentQuestionsOutput = priorOutputs?.find(
    (o) => o.section_type === 'assessment_questions'
  )

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const passageContent = assessmentPassageOutput?.content as any
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const questionsContent = assessmentQuestionsOutput?.content as any

  const passageTitle: string = passageContent?.passage_title ?? 'Assessment Passage'
  const questionsList: string = questionsContent?.questions
    ? // eslint-disable-next-line @typescript-eslint/no-explicit-any
      questionsContent.questions.map((q: any) => `${q.number}. ${q.question}`).join('\n')
    : '[Assessment questions not yet generated]'

  return `Generate a step-up writing prompt for the assessment section.

## Blueprint Context
- Lesson Title: ${blueprint.title}
- Core Concept: ${blueprint.core_concept}
- Essential Question: ${blueprint.essential_question}
- Subject: ${lesson_metadata.subject}
- Grade Level: ${lesson_metadata.grade_level}

## Assessment Passage: "${passageTitle}"
## Assessment Questions Already Given:
${questionsList}

## Instructions
Write a step-up writing prompt (2–3 sentences) that:
1. Goes beyond the comprehension questions — it asks students to synthesise, evaluate, or create
2. Requires students to use evidence from the passage
3. Ends with a clear directive to use textual evidence (e.g. "Use specific details from the text to support your response.")
4. Is ambitious but achievable for Grade ${lesson_metadata.grade_level}

## Output Schema
{
  "prompt": "string (2–3 sentences, ends with directive to use evidence)"
}

Section spec: ${JSON.stringify(spec)}`
}

File: backend/study-guide-agent/app/prompts/templates/step_up.py
from __future__ import annotations

from typing import Any

from app.types import Blueprint, GenerateRequest


def build_prompt(
    spec: dict[str, Any], blueprint: Blueprint, request: GenerateRequest
) -> str:
    assessment_passage = spec.get("assessment_passage", {})
    assessment_questions = spec.get("assessment_questions", {})

    passage_text = "\n".join(assessment_passage.get("passage", []))
    question_lines = "\n".join(
        f"- Q{item.get('number', '?')}: {item.get('question', '')}"
        for item in assessment_questions.get("questions", [])
    )

    step_up_schema = """{
    "title": "Step Up",
    "challenge_prompt": "string",
    "required_evidence": [
        "string"
    ],
    "success_criteria": [
        "string"
    ]
}"""

    prompt_lines = [
        "Create the step-up section for a K-12 study guide.",
        f"- Lesson title: {blueprint.title}",
        f"- Essential question: {blueprint.essential_question}",
        f"- Assessment passage title: {assessment_passage.get('title', 'Assessment Passage')}",
        "- Assessment passage text:",
        passage_text,
        "- Upstream assessment questions:",
        question_lines or "- none provided",
        "Requirements:",
        "- Write one higher-order challenge_prompt that builds on the assessment questions.",
        "- Make the prompt require evidence from the assessment passage rather than opinion alone.",
        "- Include required_evidence and success_criteria lists that make expectations explicit.",
        f"- Keep the reading demand close to Grade {request.lesson_metadata.grade_level}.",
        "Expected JSON schema:",
        step_up_schema,
        "Return only JSON.",
    ]

    return "\n".join(prompt_lines)
